src/SubHask/Algebra/Vector.hs

```diff
@@ -1497,6 +1515,130 @@ instance
 
 --------------------------------------------------------------------------------
 
+-- | Accelerate based Vector
+-- | A.Acc is an accelreate computation, A.Array A.DIM1 a is a one dimensional array
+
+newtype ACCVector (bknd::Backend) (n::k) a = ACCVector (A.Acc (A.Array A.DIM1 a))
+
+type instance Scalar (ACCVector bknd n r) = Scalar r
+type instance Logic (ACCVector bknd n r) = Logic r
+
+type instance ACCVector bknd m a >< b = Tensor_ACCVector (ACCVector bknd m a) b
+type family Tensor_ACCVector a b where
+    Tensor_ACCVector (ACCVector bknd n r1) (ACCVector bknd m r2) = ACCVector bknd n r1 +> ACCVector bknd m r2
+    Tensor_ACCVector (ACCVector bknd n r1) r1 = ACCVector bknd n r1 -- (r1><r2)
+
+type ValidACCVector bknd n a = ( (ACCVector (bknd::Backend) n a><Scalar a)~ACCVector (bknd::Backend) n a, Prim a, A.Elt a, A.IsNum a, IsScalar a)
+
+type instance Index (ACCVector bknd n r) = Int
+type instance Elem (ACCVector bknd n r) = Scalar r
+type instance SetElem (ACCVector (bknd::Backend) n r) b = ACCVector (bknd::Backend) n b
+
+
+type instance Actor (ACCVector (bknd::Backend) n r) = Actor r
+
+
+mkAccVectorFromList :: A.Elt a => [a] -> ACCVector bknd (n::Symbol) a
+mkAccVectorFromList l = let
+    len = P.length l
+  in ACCVector (A.use (A.fromList (A.Z A.:.len) l))
+
+mkAccVector :: (A.Elt a, ValidSVector (n::Symbol) a) => SVector (n::Symbol) a -> ACCVector (bknd::Backend) (n::Symbol) a
+mkAccVector v @(SVector_Dynamic fp off n) = let
+  arr = A.fromList (A.Z A.:. n) $ unsafeInlineIO $ go (n-1) []
+  go (-1) xs = return $ xs
+  go i    xs = withForeignPtr fp $ \p -> do
+      x <- peekElemOff p (off+i)
+      go (i-1) (x:xs)
+  in ACCVector (A.use arr)
+
+
+-- acc2SVector fails with:
+-- Could not deduce (Scalar a1 ~ Scalar (Scalar a1))
+-- from the context (ValidACCVector n a)
+--   bound by the type signature for
+--              acc2SVector :: ValidACCVector n a =>
+--                             Backend -> ACCVector n a -> SVector n a
+
+-- acc2SVector :: ValidACCVector (n::Symbol) a => Backend -> ACCVector (n::Symbol) a  -> SVector (n::Symbol) a
+-- acc2SVector bknd v = unsafeToModule (runAccVector bknd v) :: SVector (n::Symbol) a
+
+
+class ValidBackend (bknd::Backend) where
+    -- runAccVector :: (ValidACCVector bknd n a, IsScalar a) => ACCVector (bknd::Backend) n a -> SVector n a
+    runAccVector :: (ValidACCVector bknd n a, IsScalar a) => ACCVector (bknd::Backend) n a -> [a]
+
+instance ValidBackend Interpreter where
+    -- runAccVector (ACCVector a) = unsafeToModule (A.toList (I.run a)) :: SVector n a
+    runAccVector (ACCVector a) =  A.toList (I.run a)
+
+instance ValidBackend CUDA where
+    -- runAccVector (ACCVector a) = unsafeToModule (A.toList (CUDA.run a)) :: SVector n a
+    runAccVector (ACCVector a) = A.toList (CUDA.run a)
+
+-- we need an is mutable instance even though Acc types are not mutable, how to handle this?
+instance Prim a => IsMutable (ACCVector (bknd::Backend) (n::Symbol) a)
```

Comment:
This is a good question.  Everything can be given a default `IsMutable` instance with template Haskell.  This uses a `PrimRef` for mutability.  Essentially, this just fakes mutability.  I suspect if you were using raw CUDA/OpenCL it would make sense to use real mutability, but the way accelerate works, these instances are just building up a computation graph, so the fake mutability is good enough.

See:
https://github.com/mikeizbicki/subhask/blob/master/src/SubHask/Mutable.hs#L144
https://github.com/mikeizbicki/subhask/blob/master/src/SubHask/TemplateHaskell/Mutable.hs#L36